### PR TITLE
Request n

### LINKED ISSUE #229: missing backpressure exception (across TCP connection)
When streaming a response at maximum speed and using `observeOn` on the client side a MissingBackpressureException is observed. For a runnable demonstration please see https://gist.github.com/rkuhn/e450772db7ecc0c52be7e3c78d528973

### PR DESCRIPTION
**Pull request by robertroeser**
Refactored to use more of Reactor-core fluent api and removed custom code.  Removed un-used sub-module projects. Replaced Event Publisher with a Plugin class that lets you intercept, and wrap DuplexConnection, and the Client and Server ReactiveSockets. 

Fixes #229 and #237.

### CODE REVIEW

File: reactivesocket-transport-netty/src/test/java/com/rolandkuhn/rs/Client.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
+package com.rolandkuhn.rs;
```
Comment (yschimke) [inline reactivesocket-transport-netty/src/test/java/com/rolandkuhn/rs/Client.java]: what's the status and copyright of these files?

**Comment by yschimke** [comment]
I tested this branch with the cli project and I no longer get the terminal signal for streams

```
$ cat > /tmp/a.txt
a
b
c
d
$ ./reactivesocket-cli -i @/tmp/a.txt --server tcp://localhost:9898
Client
Client


$ ./reactivesocket-cli --debug --str -i Client tcp://localhost:9898
...
08:42:13.078	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] WRITE: 57B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 00 00 36 00 00 00 00 04 20 00 01 00 00 7f ff ff |..6..... .......|
|00000010| ff 7f ff ff fd 10 61 70 70 6c 69 63 61 74 69 6f |......applicatio|
|00000020| 6e 2f 6a 73 6f 6e 12 61 70 70 6c 69 63 61 74 69 |n/json.applicati|
|00000030| 6f 6e 2f 62 69 6e 61 72 79                      |on/binary       |
+--------+-------------------------------------------------+----------------+
08:42:13.079	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] FLUSH
08:42:13.086	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] Subscribing inbound receiver [pending: 0, cancelled:false, inboundDone: false]
request n from netty -> 32
08:42:13.115	sending -> Frame[0] => Stream ID: 1 Type: REQUEST_STREAM Initial Request N: 1 Payload: data: "Client"
08:42:13.116	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] Writing object PooledUnsafeDirectByteBuf(ridx: 0, widx: 19, cap: 19)
08:42:13.116	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] WRITE: 19B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 00 00 10 00 00 00 01 18 00 00 00 00 01 43 6c 69 |.............Cli|
|00000010| 65 6e 74                                        |ent             |
+--------+-------------------------------------------------+----------------+
08:42:13.117	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] FLUSH
08:42:13.121	sending -> Frame[0] => Stream ID: 1 Type: REQUEST_N RequestN: 2147483647 Payload:
08:42:13.121	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] Writing object PooledUnsafeDirectByteBuf(ridx: 0, widx: 13, cap: 13)
08:42:13.122	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] WRITE: 13B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 00 00 0a 00 00 00 01 20 00 7f ff ff ff          |....... .....   |
+--------+-------------------------------------------------+----------------+
08:42:13.122	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] FLUSH
08:42:13.124	[id: 0x47cc4050, L:/127.0.0.1:61209 - R:localhost/127.0.0.1:9898] RECEIVED: 40B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 00 00 07 00 00 00 01 28 20 61 00 00 07 00 00 00 |.......( a......|
|00000010| 01 28 20 62 00 00 07 00 00 00 01 28 20 63 00 00 |.( b.......( c..|
|00000020| 07 00 00 00 01 28 20 64                         |.....( d        |
+--------+-------------------------------------------------+----------------+
08:42:13.132	receiving -> Frame[0] => Stream ID: 1 Type: NEXT Payload: data: "a"
a
08:42:13.132	receiving -> Frame[0] => Stream ID: 1 Type: NEXT Payload: data: "b"
b
08:42:13.133	receiving -> Frame[0] => Stream ID: 1 Type: NEXT Payload: data: "c"
c
08:42:13.133	receiving -> Frame[0] => Stream ID: 1 Type: NEXT Payload: data: "d"
d
```

File: reactivesocket-core/src/main/java/io/reactivesocket/ServerReactiveSocket.java [hunk 2]
```diff
@@ -144,169 +197,106 @@ public ServerReactiveSocket start() {
...
+            UnicastProcessor<Frame> frames = UnicastProcessor.create();
+            Flux<Payload> payloads = frames
+                .doOnCancel(() -> {
+                    if (connection.availability() > 0.0) {
+                        connection.sendOne(Frame.Cancel.from(streamId)).subscribe(null, errorConsumer::accept);
+                    }
+                })
+                .doOnError(t -> {
+                    if (connection.availability() > 0.0) {
+                        connection.sendOne(Frame.Error.from(streamId, t)).doOnError(throwable -> System.out.println("EREREERERERERER")).subscribe(null, errorConsumer::accept);
```
Comment (yschimke) [inline reactivesocket-core/src/main/java/io/reactivesocket/ServerReactiveSocket.java]: test code: EREREERERERERER

File: reactivesocket-core/src/main/java/io/reactivesocket/ClientReactiveSocket.java [hunk 3]
```diff
@@ -127,70 +116,136 @@ public double availability() {
...
+                    .transform(f -> {
+                        LimitableRequestPublisher<Payload> wrapped = LimitableRequestPublisher.wrap(f);
+                        synchronized (ClientReactiveSocket.this) {
+                            senders.put(streamId, wrapped);
+                            receivers.put(streamId, receiver);
+                        }
+
+                        return wrapped;
+                    })
+                    .doOnRequest(l -> System.out.println("request n from netty -> "  + l))
```
Comment (yschimke) [inline reactivesocket-core/src/main/java/io/reactivesocket/ClientReactiveSocket.java]: test code

**Comment by yschimke** [comment]
I haven't debugged why it has changed, but I also don't get the exception message delivered to the client anymore

```
$ reactivesocket-cli --debug -i @/tmp/assss.txt --server tcp://localhost:9899
...
09:23:20.419	an error occurred, creating error frame
io.reactivesocket.cli.UsageException: file not found: /tmp/assss.txt
	at io.reactivesocket.cli.Main.inputFile(Main.java:365)
...
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)

09:23:20.421	Frame[0] => Stream ID: 1 Type: ERROR Error code: 516 Payload: data: "file not found: /tmp/assss.txt"
09:23:20.422	[id: 0x899f7d10, L:/127.0.0.1:9899 - R:/127.0.0.1:53399] Writing object PooledUnsafeDirectByteBuf(ridx: 0, widx: 43, cap: 43)
09:23:20.423	[id: 0x899f7d10, L:/127.0.0.1:9899 - R:/127.0.0.1:53399] WRITE: 43B
         +-------------------------------------------------+
         |  0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f |
+--------+-------------------------------------------------+----------------+
|00000000| 00 00 2b 00 00 00 01 2c 00 00 00 02 04 66 69 6c |..+....,.....fil|
|00000010| 65 20 6e 6f 74 20 66 6f 75 6e 64 3a 20 2f 74 6d |e not found: /tm|
|00000020| 70 2f 61 73 73 73 73 2e 74 78 74                |p/assss.txt     |
+--------+-------------------------------------------------+----------------+
```

On this branch

```
io.reactivesocket.cli.UsageException: file not found: /tmp/assss.txt
	at io.reactivesocket.cli.Main.inputFile(Main.java:355)
	at io.reactivesocket.cli.Main.inputPublisher(Main.java:323)
	at io.reactivesocket.cli.Main.handleIncomingPayload(Main.java:266)
	at io.reactivesocket.cli.Main.access$100(Main.java:68)
	at io.reactivesocket.cli.Main$1.requestStream(Main.java:247)
	at io.reactivesocket.util.ReactiveSocketProxy.requestStream(ReactiveSocketProxy.java:47)
	at io.reactivesocket.ServerReactiveSocket.requestStream(ServerReactiveSocket.java:95)
	at io.reactivesocket.ServerReactiveSocket.lambda$start$2(ServerReactiveSocket.java:140)
	at reactor.core.publisher.FluxFlatMap$FlatMapMain.onNext(FluxFlatMap.java:332)
	at reactor.core.publisher.MonoFlatMap$FlattenSubscriber$InnerSubscriber.onNext(MonoFlatMap.java:199)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onNext(FluxPeekFuseable.java:178)
	at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.drainRegular(FluxGroupBy.java:592)
	at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.drain(FluxGroupBy.java:668)
	at reactor.core.publisher.FluxGroupBy$UnicastGroupedFlux.subscribe(FluxGroupBy.java:733)
	at reactor.core.publisher.FluxPeekFuseable.subscribe(FluxPeekFuseable.java:87)
	at reactor.core.publisher.MonoFlatMap$FlattenSubscriber.onNext(MonoFlatMap.java:161)
	at reactor.core.publisher.ReplayProcessor$SizeBoundReplayBuffer.replayNormal(ReplayProcessor.java:857)
	at reactor.core.publisher.ReplayProcessor$SizeBoundReplayBuffer.replay(ReplayProcessor.java:941)
	at reactor.core.publisher.ReplayProcessor.onNext(ReplayProcessor.java:437)
	at reactor.core.publisher.MonoProcessor.drainLoop(MonoProcessor.java:483)
	at reactor.core.publisher.MonoProcessor.onNext(MonoProcessor.java:322)
	at io.reactivesocket.internal.ClientServerInputMultiplexer.lambda$new$1(ClientServerInputMultiplexer.java:81)
	at reactor.core.publisher.LambdaSubscriber.onNext(LambdaSubscriber.java:139)
	at reactor.core.publisher.FluxGroupBy$GroupByMain.drainLoop(FluxGroupBy.java:422)
	at reactor.core.publisher.FluxGroupBy$GroupByMain.drain(FluxGroupBy.java:358)
	at reactor.core.publisher.FluxGroupBy$GroupByMain.onNext(FluxGroupBy.java:208)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	at reactor.core.publisher.FluxMap$MapSubscriber.onNext(FluxMap.java:114)
	at reactor.ipc.netty.channel.FluxReceive.drainReceiver(FluxReceive.java:204)
	at reactor.ipc.netty.channel.FluxReceive.onInboundNext(FluxReceive.java:319)
	at reactor.ipc.netty.channel.ChannelOperations.onInboundNext(ChannelOperations.java:312)
	at reactor.ipc.netty.channel.ChannelOperationsHandler.channelRead(ChannelOperationsHandler.java:125)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:293)
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:280)
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:396)
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:248)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:233)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:341)
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1334)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:363)
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:349)
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:926)
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:129)
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:642)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:565)
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:479)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:441)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
	at java.lang.Thread.run(Thread.java:745)
09:36:21.215	An exception 'reactor.core.Exceptions$CancelException: The subscriber has denied dispatching' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
reactor.core.Exceptions$CancelException: The subscriber has denied dispatching
	at reactor.core.Exceptions.failWithCancel(Exceptions.java:129)
	at reactor.core.publisher.Operators.onNextDropped(Operators.java:306)
	at reactor.core.publisher.MonoProcessor.onNext(MonoProcessor.java:289)
```

**Comment by yschimke** [comment]
Apologies, I didn't mean to close.  Not sure how it happened? maybe close and comment instead of comment.

**Comment by yschimke** [comment]
BTW I've put changes to use reactivesocket-cli with this branch here https://github.com/yschimke/reactivesocket-cli/tree/branch_test

you can run from a checkout with

$ ./reactivesocket-cli

**Comment by robertroeser** [comment]
@yschimke committed a fix - you can try the client again.

**Comment by yschimke** [comment]
Thanks, I get the terminal signal and I get error frames.  

However if I kill the server while processing streams the client (over TCP) doesn't get an error, just hangs as my client has a blockingGet. I believe that is also new.

**Comment by yschimke** [comment]
LGTM - I think there may be some cleanup with how connection drops are handled, but I'm happy with this change functionally.  Probably easier to make incremental changes after this lands.

```
10:54:55.121	An exception 'reactor.core.Exceptions$ErrorCallbackNotImplemented: java.io.IOException: Connection reset by peer' [enable DEBUG level for full stacktrace] was thrown by a user handler's exceptionCaught() method while handling the following exception:
java.io.IOException: Connection reset by peer
	at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
	at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
	at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
```

File: reactivesocket-transport-local/src/main/java/io/reactivesocket/transport/local/LocalClient.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package io.reactivesocket.transport.local;
```
Comment (somasun) [inline reactivesocket-transport-local/src/main/java/io/reactivesocket/transport/local/LocalClient.java]: AbstractReactiveSocketPerf seems to reference this as "io.reactivesocket.local".   Should we change the import there?  Unless I am missing something.

https://github.com/ReactiveSocket/reactivesocket-java/blob/1.0.x/reactivesocket-examples/src/jmh/java/io/reactivesocket/perf/AbstractReactiveSocketPerf.java

**Pull request merged**